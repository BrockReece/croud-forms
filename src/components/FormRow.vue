<template>
    <tr>
        <td class="wide top aligned" :class="labelClasses">{{{ field.name }}}</td>
        <td class="wide" :class="inputClasses">
            <component :is="view" :model.sync="model" :field.sync="field" :transparent="transparent" :read-only="readOnly"></component>
        </td>
    </tr>
</template>

<script>
    import Text from './fields/Text.vue'
    import Number from './fields/Number.vue'
    import MultiCheckbox from './fields/MultiCheckbox.vue'
    import Textarea from './fields/Textarea.vue'
    import CroudSelect from './fields/Select.vue'
    import SearchSelect from './fields/SearchSelect.vue'
    import CroudRadio from './fields/Radio.vue'
    import SortCode from './fields/SortCode.vue'
    import General from './fields/General.vue'

    export default {
        props: {
            model: {
                required: true,
                twoWay: true,
            },
            field: {
                required: true,
                twoWay: true,
            },
            readOnly: {
                default: false,
            },
            columnWidths: {
                default() {
                    return {
                        label: 'nine',
                        input: 'severn',
                    }
                },
            },
            transparent: {
                default: true,
            },
        },

        computed: {
            labelClasses() {
                const classes = {}
                classes[this.columnWidths.label] = true

                return classes
            },
            inputClasses() {
                const classes = {}
                classes[this.columnWidths.input] = true

                return classes
            },
            view() {
                if (['select', 'radio'].indexOf(this.field.field_type) !== -1) {
                    return `croud-${this.field.field_type}`
                }
                if (['search-select', 'text', 'number', 'multi-checkbox', 'textarea', 'sort-code'].indexOf(this.field.field_type) !== -1) {
                    return this.field.field_type
                }
                return 'general'
            },
        },

        components: {
            Number,
            Text,
            Textarea,
            MultiCheckbox,
            CroudSelect,
            CroudRadio,
            General,
            SortCode,
            SearchSelect,
        },
    }
</script>
