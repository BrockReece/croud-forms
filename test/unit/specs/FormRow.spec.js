import { mount } from 'vue-test-utils'
import FormRow from '@/components/FormRow'

const fieldTypes = ['text', 'number', 'checkbox']

describe('FormField component', () => {
    describe('should render correct field type', () => {
        fieldTypes.forEach((type) => {
            it(type, () => {
                const vm = mount(FormRow, {
                    propsData: {
                        field: {
                            field_name: 'First name',
                            field_type: type,
                        },
                        model: 'test',
                    },
                })
                expect(vm.html()).toMatchSnapshot()
            })
        })
    })

    it('should render a non transparent input', () => {
        const vm = mount(FormRow, {
            propsData: {
                field: {
                    field_name: 'First name',
                    field_type: 'text',
                },
                model: 'test',
                transparent: false,
            },
        })
        expect(vm.html()).toMatchSnapshot()
    })
})
