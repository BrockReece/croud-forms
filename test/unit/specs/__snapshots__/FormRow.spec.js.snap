// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormField component should render a non transparent input 1`] = `
<tr>
    <td class="wide top aligned nine">First name</td>
    <td class="wide severn">
        <div class="ui fluid input"><input type="text" placeholder="First name"></div>
    </td>
</tr>
`;

exports[`FormField component should render correct field type checkbox 1`] = `
<tr>
    <td class="wide top aligned nine">First name</td>
    <td class="wide severn">
        <div class="ui input transparent">
            <semantic-checkbox type="checkbox"></semantic-checkbox>
        </div>
    </td>
</tr>
`;

exports[`FormField component should render correct field type number 1`] = `
<tr>
    <td class="wide top aligned nine">First name</td>
    <td class="wide severn">
        <div class="ui input transparent"><input type="number" placeholder="First name"></div>
    </td>
</tr>
`;

exports[`FormField component should render correct field type text 1`] = `
<tr>
    <td class="wide top aligned nine">First name</td>
    <td class="wide severn">
        <div class="ui fluid input transparent"><input type="text" placeholder="First name"></div>
    </td>
</tr>
`;
